### PR TITLE
[kayabuttertoastt] iP

### PR DESCRIPTION
**Pull request by kayabuttertoastt**
# ChatBot 

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

## Features 
* Add Todo
* Add Deadlines 
* Add Events 
* Delete Tasks
* Mark/Unmark Tasks 

## Getting Started 
1. Download it from [here](https://nus-cs2103-ay2324s1.github.io/website/admin/ip-w4.html).
2. Double-click it.
3. Add your tasks.

And the best part is, its **FREE**!! 🤩

## Walk Through
To add `todo`, 
```
todo run
``` 
To add `deadline`, 
```
deadline email /by 11-11-2023 19:00:00
```
To add `event`, 
```
event sports day /from 11-11-2023 08:00:00 /to 12-11-2023 20:00:00
```

## Future Improvements 
- [ ] Improve GUI

### CODE REVIEW

File: src/main/java/chatbot/commands/Command.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    public abstract boolean isExit();
+
+    /**
+     * Abstract method that executes the command that adds the task
+     * @param tasksList Task list which contains the tasks
+     * @param ui A UI instance that displays a message to indicate to the user the task has been added
+     * @param storage Storage instance that represents the storage of the file
+     */
+    public abstract void execute(TaskList tasksList, Ui ui, Storage storage);
+}
```
Comment (xavierpok) [inline src/main/java/chatbot/commands/Command.java]: I like the nice usage of inheritance here, seems to make it much more extensible!

File: src/main/java/chatbot/parser/Parser.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                && !command.contains("/from") && !command.contains("/to")) {
+            throw new DukeException("Error! Input should have a description, start date and end date");
+        } else if(command.split(" ")[0].equals("mark") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the task to be marked");
+        } else if(command.split(" ")[0].equals("unmark") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the task to be unmarked");
+        } else {
+            return true;
+        }
+    }
```
Comment (xavierpok) [inline src/main/java/chatbot/parser/Parser.java]: Maybe this could be delegated to the Command classes to figure out what kind of syntax they should accept?

Granted, implementation might get a bit messy, and this certainly works!

File: src/main/java/chatbot/commands/DeadlineCommand.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+public class DeadlineCommand extends Command {
+    /**
+     * Description of the deadline to be added
+     */
+    String description;
+
+    /**
+     * Due date for the task
+     */
+    String date;
```
Comment (xavierpok) [inline src/main/java/chatbot/commands/DeadlineCommand.java]: Minor issue, but these fields are lacking access modifiers. Maybe they should be set to private?

```suggestion
    /**
     * Description of the deadline to be added
     */
    private String description;

    /**
     * Due date for the task
     */
    private String date;
```

File: src/main/java/chatbot/commands/EventCommand.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Start date of the event
+     */
+    String from;
+
+    /**
+     * End date of the event
+     */
+    String to;
```
Comment (xavierpok) [inline src/main/java/chatbot/commands/EventCommand.java]: This might need access modifiers to be more explicit
```suggestion
    /**
     * Description of the event task to be added
     */
    private String description;

    /**
     * Start date of the event
     */
    private String from;

    /**
     * End date of the event
     */
   private  String to;
```

File: src/main/java/chatbot/tasks/Event.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    protected String start;
+
+    /**
+     * Formatted start date of event task
+     */
+    protected LocalDateTime formattedStart;
+    /**
+     * Formatted end date of event task
+     */
+    protected LocalDateTime formattedEnd;
```
Comment (xavierpok) [inline src/main/java/chatbot/tasks/Event.java]: I'm not entirely sure why these are currently `protected`, maybe they could be set to `private`?
```suggestion
    /**
     * End date of event task
     */
    private String end;
    /**
     * Start date of event task
     */
    private String start;

    /**
     * Formatted start date of event task
     */
    private LocalDateTime formattedStart;
    /**
     * Formatted end date of event task
     */
    private LocalDateTime formattedEnd;
```

**Review by xavierpok — COMMENTED** [review]
Overall, LGTM, just a few nits.

File: src/main/java/chatbot/parser/Parser.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        } else if(command.split(" ")[0].equals("event") && command.split(" ")[1].length() < 1
+                && !command.contains("/from") && !command.contains("/to")) {
+            throw new DukeException("Error! Input should have a description, start date and end date");
+        } else if(command.split(" ")[0].equals("mark") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the task to be marked");
+        } else if(command.split(" ")[0].equals("unmark") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the task to be unmarked");
+        } else {
+            return true;
+        }
```
Comment (ItsTYtan) [inline src/main/java/chatbot/parser/Parser.java]: Could perhaps use a switch statement and split the string command at the top, since some splits overlap in the code. I think the validation of the command can be done at the same time as parsing the command and deciding which command to execute.

File: src/main/java/chatbot/tasks/Deadline.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     * Returns the formatted string representation of the deadline
+     * @return Formatted string representation of the deadline
+     */
+    @Override
+    public String store() {
+        return String.format("D | %s | %s | %s", this.isDone, this.description,
+                formattedDate.format(DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss")));
+    }
```
Comment (ItsTYtan) [inline src/main/java/chatbot/tasks/Deadline.java]: Nice use of formatters

File: src/main/java/chatbot/tasks/TaskList.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+     */
+    public void removeTask(int num) {
+        if(num < 0){
+            System.out.println("The value cannot be negative. Please try again.");
+        } else if (num >= this.inputs.size()){
+            System.out.println("The input value is wrong. Please try again.");
+        } else {
+            this.inputs.remove(num);
+        }
+    }
```
Comment (ItsTYtan) [inline src/main/java/chatbot/tasks/TaskList.java]: Maybe have an exception for invalid nums?

File: src/main/java/chatbot/tasks/TaskList.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        return this.inputs.size();
+    }
+
+    /**
+     * Retrieves the list of tasks stored in the task list
+     * @return ArrayList that contains the tasks stored in the task list
+     */
+    public ArrayList<Task> retrieveList() {
+        return this.inputs;
+    }
```
Comment (ItsTYtan) [inline src/main/java/chatbot/tasks/TaskList.java]: Not sure if its a good idea to retrieve the arraylist in the tasklist class itself. I think any operation on the tasklist can be done without exposing the inner mutatable array

**Review by ItsTYtan — COMMENTED** [review]
LGTM, just some nitpicking on minor issues.

File: src/main/java/chatbot/commands/DeadlineCommand.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+     * @param description Description of the deadline to be added
+     * @param date Due date for the task
+     */
+    public DeadlineCommand(String description, String date) {
+        this.description = description;
+        this.date = date;
+    }
+
+    /**
+     *  Returns a boolean value to indicate whether to exit the program
```
Comment (nicrandomlee) [inline src/main/java/chatbot/commands/DeadlineCommand.java]: Extra spacing in javadoc

File: src/main/java/chatbot/parser/Parser.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+     * Parses the user input and returns the corresponding command
+     * @param command The user input to be parsed.
+     * @return Command that corresponds to the parsed user input
+     * @throws DukeException If the user input is not valid
+     */
+    public static Command parse(String command) throws DukeException {
+
+        isValid(command);
+
+        if (command.equals("bye")) {
```
Comment (nicrandomlee) [inline src/main/java/chatbot/parser/Parser.java]: You can try doing a switch statement to make the code more readable. Feel free to use enums to represent commands as well so it is clearer.

Eg:
case BYE:
<insert code here>

File: src/main/java/chatbot/parser/Parser.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+                && !command.contains("/by")) {
+            throw new DukeException("Error! Input should have a description and date");
+        } else if (command.split(" ")[0].equals("event") && command.split(" ")[1].length() < 1
+                && !command.contains("/from") && !command.contains("/to")) {
+            throw new DukeException("Error! Input should have a description, start date and end date");
+        } else if (command.split(" ")[0].equals("mark") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the task to be marked");
+        } else if (command.split(" ")[0].equals("unmark") && command.split(" ")[1].length() < 1) {
+            throw new DukeException("Error! Please include a description of the task to be unmarked");
+        } else {
```
Comment (nicrandomlee) [inline src/main/java/chatbot/parser/Parser.java]: Just nitpicking a little but you can consider adding the format that the user is supposed to enter so that they have a point of reference, since most of the error messages are generic for the particular task entered and the user may not know what went wrong.

File: src/main/java/chatbot/tasks/TaskList.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+        this.inputs.add(task);
+    }
+
+    /**
+     * Removes the task at a particular index as done
+     * @param num Index of the particular task in the task list
+     */
+    public void removeTask(int num) {
+        if (num < 0) {
+            System.out.println("The value cannot be negative. Please try again.");
```
Comment (nicrandomlee) [inline src/main/java/chatbot/tasks/TaskList.java]: Good handling of edge cases!

**Review by nicrandomlee — COMMENTED** [review]
Overall, I found your code really easy to read. Good use of OOP to handle various commands and coding standards too! I am very thankful to have reviewed your code as I learnt a lot from your code writing style. Keep it up!

File: src/main/java/chatbot/commands/DeadlineCommand.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+import chatbot.tasks.Deadline;
+import chatbot.tasks.Task;
+import chatbot.tasks.TaskList;
+import chatbot.ui.Ui;
+
+/**
+ * Represents a command to add a deadline to the task list
+ */
+public class DeadlineCommand extends Command {
+    /**
```
Comment (Cleon2) [inline src/main/java/chatbot/commands/DeadlineCommand.java]: Need to add spacing here for the javadoc

File: src/main/java/chatbot/commands/TodoCommand.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+     * Description of the todo task to be added
+     */
+    String description;
+
+    /**
+     * Constructor for TodoCommand
+     * @param description Description of the todo task to be added
+     */
+    public TodoCommand(String description) {
+
```
Comment (Cleon2) [inline src/main/java/chatbot/commands/TodoCommand.java]: unnecessary spacing here

File: src/main/java/chatbot/commands/Command.java [hunk 16]
```diff
@@ -0,0 +1,24 @@
+package chatbot.commands;
+
+import chatbot.storage.Storage;
+import chatbot.tasks.TaskList;
+import chatbot.ui.Ui;
+
+/**
+ * Represents a command that can be executed by the chatbot
+ */
+public abstract class Command {
```
Comment (Cleon2) [inline src/main/java/chatbot/commands/Command.java]: Love how you abstracted everything using command!

File: src/main/java/chatbot/parser/Parser.java [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+ * Represents the class that is responsible for parsing user inputs and generating the relevant commands
+ */
+public class Parser {
+    /**
+     * Checks if the user input is valid
+     * @param command User input to be validated
+     * @return Returns a boolean that indicates if the input is valid
+     * @throws DukeException If the user input is not valid
+     */
+    public static boolean isValid (String command) throws DukeException {
```
Comment (Cleon2) [inline src/main/java/chatbot/parser/Parser.java]: Code seems a bit wordy here

**Review by Cleon2 — COMMENTED** [review]
Overall looks good o
